macaw-aarch32/src/Data/Macaw/ARM/Arch.hs

```diff
@@ -96,26 +96,59 @@ data ARMBlockPrecond =
 -- ARM terminal statements (which have instruction-specific effects on
 -- control-flow and register state).
 
-data ARMTermStmt ids where
-  ARMSyscall :: WI.W 24 -> ARMTermStmt ids
-  ThumbSyscall :: WI.W 8 -> ARMTermStmt ids
-
-deriving instance Show (ARMTermStmt ids)
+data ARMTermStmt f where
+  ARMSyscall :: WI.W 24 -> ARMTermStmt f
+  ThumbSyscall :: WI.W 8 -> ARMTermStmt f
+  -- | Return if the condition is true; otherwise, fall through to the next instruction
+  ReturnIf :: f MT.BoolType -> ARMTermStmt f
+  -- | Return if the condition is not true; otherwise, fall through to the next instruction
+  --
+  -- Note that it is unfortunate that we need this in addition to 'ReturnIf'. At
+  -- the part of the analysis where we are able to generate these block
+  -- terminators, we cannot generate new statements since we do not have the
+  -- right nonce generator. As a result, if we are returning when the condition
+  -- is false, we are not able to generate a negation (which would require
+  -- wrapping the original condition in an Assignment, which has an AssignId,
+  -- which requires a nonce). We work around this by just having an additional
+  -- block terminator.
+  ReturnIfNot :: f MT.BoolType -> ARMTermStmt f
```

Comment:
If you combined these to `ReturnCond :: Bool -> f MT.BoolType -> ARMTermStmt f` (or even better, `ReturnCond :: ReturnsOnBranch -> f MT.BoolType -> ARMTermStmt f` if you moved `ReturnsOnBranch` to be visible here) then you'd only have a single constructor, which would save a lot of code duplication.  It doesn't look like the different constructors provides any pattern-matching value other than in `aarch32GenTermStmt`, which could just as easily match on `ReturnsOnBranch` constructors.